# MODA-Wrapper
#### MODA <small>-</small> Module Differential Analysis
>for weighted gene co-expression network analysis

## Description
MODA can be used to estimate and construct condition-specific gene co-expression networks, and identify differentially expressed subnetworks as conserved or condition specific modules which are potentially associated with relevant biological processes

## Features

With __MODA-Wrapper__, you can perform the following: 
1. Hierarchical clustering
2. Community detection
3. Spectral clustering
4. **Network comparison**


## Installation Instructions

There are two ways to quickly set up MODA-Wrapper.
#### Alternative 1) R Console:
##### Dependencies
> - R version >= **4.3**
> - BiocManager

Step 1:
* MODA dependencies</br>
Installation of MODA requires prior installation of BiocManager:
```R
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
```
Step 2:
- Install MODA:
```R
BiocManager::install("MODA")
```

```R
install.packages(data.table)
```

- To test that MODA installed correctly execute:

```bash
library(MODA)
```

Step 3:
* Get **MODA-Wrapper** repository:

```shell
git clone https://github.com/bionetslab/grn-benchmark
cd src/moda-wrapper/R/
```

#### Alternative 2) Docker:
By using Docker, all necessary dependencies comes with the Docker image. Therefore, first clone the repository and navigate to the `docker` sub-directory:

```shell
git clone https://github.com/bionetslab/grn-benchmark
cd src/moda-wrapper/docker/
```

##### Build Image
A <font color="#f0ad4e"> dockerfile</font> which contains all the needed build instructions is located in the MODA-Wrapper repo's docker (<font color="#f0ad4e">moda-wrapper/docker</font>) directory. Execute the docker command below to build the docker image:


```shell
mkdir output
docker build -t moda-wrapper .
```

Run the docker container:
```shell
docker run -v /host_path:/tool moda-wrapper Rscript script-name.R options
```
- `host_path` represents the directory shared between the host and Docker, additionally it is where your dataset files are located. The output files generated by MODA during the analysis will be written into this directory. 
- `script-name` is the corresponding MODA functionality you want to execute. `e.g: hierarchical-clustering.R`
 
- `options` represent the parameters available for the respective MODA-Wrapper script, as detailed below.
## 1.  Hierarchical Clustering

Detection of Modules (group of genes that co-express) by hierarchical clustering

#### Usage 

```shell
Rscript hierarchical-clustering.R ./input_file_1_path/ ./input_file_2_path/ output_path/ options 
```

#### Required Arguments


1. **Input file 1**:<font color="#d9534f">String:required</font><br>Path to a tab-separated file that contains the normalized gene expression for condition 1
    * First column is named 'Gene' and contains the gene names
    * All following columns are named after a sample/cell and contain the normalized gene expression for each respective gene for the given sample/cell

2. **Input file 2**:<font color="#d9534f">String:required</font><br>Path to a tab-separated file that contains the normalized gene expression for condition 2
    * First column is named 'Gene' and contains the gene names
    * All following columns are named after a sample/cell and contain the normalized gene expression for each respective gene for the given sample/cell

3. **Output path**:<font color="#d9534f">String:required</font><br>String of the output directory. The directory must exist **prior** to execution of the script!



#### Parameters (options)
Hierarchical clustering script accepts optional parameters as detailed below
* **-i1**
Type: `String`<br>
Description: Indicator or name for condition 1. The choosen name will form part of the gene module files generated as output 
 ```e.g: -i1 Macrophages ```

* **-i2**
Type: `String`<br>
Description: Indicator or name for condition 2. The choosen name will form part of the gene module files generated as output 
 ```e.g: -i1 TCells ```

* **-c**
  Type: `Integer`<br>
  Description: Specify the desired cutting method for modules. Accepts an integer value of either 1 or 2. Use 1 for `Density` and 2 for `Modularity`.
  Example: `-c 2`

* **-s**
  Type: `Numeric`<br>
  >Description: Threshold to define $min(s) + \theta_1$, less than which is considered as condition specific module. ```s``` is the sums of rows in Jaccard index matrix. [See Supplementary file](https://www.biorxiv.org/content/biorxiv/suppl/2016/05/16/053496.DC1/053496-1.pdf)
  ```e.g: -s 0.1 ```  
* **-t**
  Type: `Numeric`<br>
  >Description: Threshold to define $max(s)-\theta_2$, greater than which is considered as condition conserved module. ```s``` is the sums of rows in Jaccard index matrix. [See Supplementary file](https://www.biorxiv.org/content/biorxiv/suppl/2016/05/16/053496.DC1/053496-1.pdf)
  ```e.g: -t 0.1 ```




##### Exemplery Execution Instructions (Rscript):

```shell
git clone https://github.com/bionetslab/grn-benchmark
cd src/moda-wrapper/R/
```

```bash
mkdir output
```

```bash
Rscript hierarchical-clustering.R ./dataset/Macrophages/ ./dataset/TCells/ output/ -i1 Macrophages -i2 TCells -c 2 -s 0.1 -t 0.1
```


##### Exemplery Execution Instructions (Docker):

```bash
git clone https://github.com/bionetslab/grn-benchmark
cd src/moda-wrapper/docker/
```

```bash
mkdir output
docker build -t moda-wrapper .
```

```shell
docker run -v ./:/tool moda-wrapper Rscript hierarchical-clustering.R ./dataset/Macrophages/ ./dataset/TCells/ output/ -i1 Macrophages -i2 TCells -c 2 -s 0.1 -t 0.1
```

##### Interpretation of the output

After the script runs successfully, it produces a directory named `output-hierarchical`. Assuming the specified output directory for the script in the script arguments is `output`, the resulting directory structure would appear as:

```
 output
   └── output-hierarchical
       ├── DenseModuleGene_Macrophages_1.txt
       ├── DenseModuleGene_Macrophages_2.txt
                    .
                    .
                    .
       ├── DenseModuleGeneID_Macrophages_1.txt
                    .
                    .
                    .
```

 This directory contains text files, one for each identified module, containing gene names and gene ids in the identified module.
`e.g: DenseModuleGene_Macrophages_1.txt, DenseModuleGeneID_Macrophages_1.txt`

Additionally, it generates images such as dendrogan, as shown below.

![Hierarchical clustering](https://raw.githubusercontent.com/Turnyur/image-hub/main/Partitions_Macrophages.png)

![Hierarchical clustering](https://raw.githubusercontent.com/Turnyur/image-hub/main/heatdata_TCells.png)


## 2. Community Detection

Community detection using Louvain algorithm.

#### Usage 

```Shell
Rscript community-detection.R ./input_path/ output_path/
```

#### Required Arguments


1. **Input file 1**:<font color="#d9534f">String:required</font><br>
 Path to a tab-separated file that contains the normalized gene expression for condition 1
    * First column is named 'Gene' and contains the gene names
    * All following columns are named after a sample/cell and contain the normalized gene expression for each respective gene for the given sample/cell


3. **Output path**:<font color="#d9534f">String:required</font><br>
 String of the output directory. The directory must exist **prior** to execution of the script!



#### Parameters (options)
These are the  optional parameters for the script.
* **-i1**
Type: `String`<br>
Description: Indicator or name for gene data profile. The choosen name will form part of the gene module files generated as output 
 ```e.g: -i1 Macrophages ```

##### Exemplery Execution Instructions (Rscript):
```bash
Rscript community-detection.R ./dataset/Macrophages/  output/ -i1 Macrophages
```


##### Exemplery Execution Instructions (Docker):
```bash
docker run -v ./:/tool moda-wrapper  Rscript community-detection.R ./dataset/Macrophages/  output/ -i1 Macrophages
```
##### Interpretation of the output

```
output
   └── output-louvain
```
After the script runs successfully, it produces a directory named `output-louvain`
This directory contains text files of identified communities (groups of densely connected genes)
`e.g: DenseModuleGene_Macrophages_1.txt`


## 3. Spectral Clustering


#### Usage 

```Shell
Rscript hierarchical-clustering.R ./input_path/ output_path/
```

#### Required Arguments


1. **Input file 1**:<font color="#d9534f">String:required</font><br>
 Path to a tab-separated file that contains the normalized gene expression for condition 1
    * First column is named 'Gene' and contains the gene names
    * All following columns are named after a sample/cell and contain the normalized gene expression for each respective gene for the given sample/cell


3. **Output path**:<font color="#d9534f">String:required</font><br>
 String of the output directory. The directory must exist **prior** to execution of the script!



#### Parameters (options)
These are the  optional parameters for the script.
* **-i1**
Type: `String`<br>
Description: Indicator or name for gene data profile. The choosen name will form part of the gene module files generated as output 
 ```e.g: -i1 Macrophages ```

* **-k**
Type: `Integer`<br>
Description: k number of desired clusters 
 ```e.g: -k 5 ``` for five clusters



##### Exemplery Execution Instructions (Rscript):
```bash
Rscript spectral_clustering.R ./dataset/Macrophages/  output/ -i1 Macrophages k 6
```

##### Exemplery Execution Instructions (Docker):
```bash
docker run -v ./:/tool moda-wrapper  Rscript spectral_clustering.R ./dataset/Macrophages/  output/ -i1 Macrophages k 6
```
##### Interpretation of the output
```
output
   └── output-spectral
```


On successful completion of the script, it produces a directory named `output-spectral`
This directory contains text files of identified modules.
`e.g: DenseModuleGene_Macrophages_1.txt, DenseModuleGene_Macrophages_1`



## 4.  Network Comparison

 Can identify differentially expressed subnetworks as conserved or condition-specific modules.
#### Usage 

```Shell
Rscript hierarchical-clustering.R ./input_file_1_path/ ./input_file_2_path/ output_path/ 
```

#### Required Arguments


1. **Input file 1**:<font color="#d9534f">String:required</font><br>
 Path to a tab-separated file that contains the normalized gene expression for condition 1
    * First column is named 'Gene' and contains the gene names
    * All following columns are named after a sample/cell and contain the normalized gene expression for each respective gene for the given sample/cell



2. **Input file 2**:<font color="#d9534f">String:required</font><br>
 Path to a tab-separated file that contains the normalized gene expression for condition 2
    * First column is named 'Gene' and contains the gene names
    * All following columns are named after a sample/cell and contain the normalized gene expression for each respective gene for the given sample/cell

3. **Output path**:<font color="#d9534f">String:required</font><br>
 String of the output directory. The directory must exist **prior** to execution of the script!



#### Parameters (options)
These are the  optional parameters for the script.
* **-i1**
Type: `String`<br>
Description: Indicator or name for condition 1. The choosen name will form part of the gene module files generated as output 
 ```e.g: -i1 Macrophages ```

* **-i2**
Type: `String`<br>
Description: Indicator or name for condition 2. The choosen name will form part of the gene module files generated as output 
 ```e.g: -i1 TCells ```

* **-c**
  Type: `Integer`<br>
  Description: Specify the desired cutting method for modules. Accepts an integer value of either 1 or 2. Use 1 for `Density` and 2 for `Modularity`.
  Example: `-c 2`

* **-s**
  Type: `Numeric`<br>
  >Description: Threshold to define $min(s) + \theta_1$, less than which is considered as condition specific module. ```s``` is the sums of rows in Jaccard index matrix [See Supplementary file](https://www.biorxiv.org/content/biorxiv/suppl/2016/05/16/053496.DC1/053496-1.pdf).
  ```e.g: -s 0.1 ```
* **-t**
  Type: `Numeric`<br>
  >Description: Threshold to define $max(s)-\theta_2$, greater than which is considered as condition conserved module. ```s``` is the sums of rows in Jaccard index matrix [See Supplementary file](https://www.biorxiv.org/content/biorxiv/suppl/2016/05/16/053496.DC1/053496-1.pdf).
  ```e.g: -t 0.1 ```


##### Exemplery Execution Instructions (Rscript):


```bash
Rscript network-comparison.R ./dataset/Macrophages/ ./dataset/TCells/ output/ -i1 Macrophages -i2 TCells -c 1 -s 0.1 -t 0.1
```

##### Exemplery Execution Instructions (Docker):


```bash
docker run -v ./:/tool moda-wrapper Rscript network-comparison.R ./dataset/Macrophages/ ./dataset/TCells/ output/ -i1 Macrophages -i2 TCells -c 1 -s 0.1 -t 0.1
```



##### Interpretation of the output
```
 output
   └── output-network
   |    ├── DenseModuleGene_Macrophages_1.txt
   |    │             .
   |    │             .
   |    │             .
   |    ├── TCells
   |    │    ├── conservedModuleid.txt
   |    │    ├── module_overlap_removeTCells.png
   |    │    ├── specificModuleid.txt
   |    |
   ├── condition_specific_gene_list.txt
   ├── conserved_gene_list.txt
   └── network.tsv
```
The script generates a `output-network` directory that includes output files such as module clusters containing module IDs and separate files with corresponding gene names. Each identified module is represented by a text file within this directory


Additionally the script produces a sub-directory containing IDs of  modules categorized as `conservedModules` or `conditionSpecificModules`
`e.g: DenseModuleGene_Macrophages_1.txt, DenseModuleGeneID_Macrophages_1.txt`


MODA-Wrapper further post-processes the output files generated after the network comparison and creates a <font color="#f0ad4e">network.tsv</font> file.
This file contains the following columns `ModuleID`, `Gene`, `Condition`, `ClusterMethod`. It does this by using the module IDs in the secondary condition (TCell) and loads the corresponding genes in the given module. 

Additionally, MODA-Wrapper produces two files (<font color="#f0ad4e">condition_specific_gene_list.txt</font> and <font color="#f0ad4e">conserved_gene_list.txt</font>) containing the list of gene symbols which were identified as condition-specific genes and conserved genes respectively, which can further be used for Functional Annotation Analysis with tools such as [DAVID](https://david.ncifcrf.gov/summary.jsp). </br>Example Functional annotation analysis of list of genes in the <font color="#f0ad4e">condition_specific_gene_list.txt</font> file, using [DAVID](https://david.ncifcrf.gov/summary.jsp) can be found in `downstream` sub-directory within MODA-Wrapper root directory.


## :bulb: Additional resources

-    [MODA Paper](https://www.biorxiv.org/content/10.1101/053496v4)
-    [MODA Manual](https://bioconductor.org/packages/release/bioc/manuals/MODA/man/MODA.pdf)
-    [MODA package site](https://bioconductor.org/packages/release/bioc/html/MODA.html)

-    [MODA Tutorial](https://www.bioconductor.org/packages/release/bioc/vignettes/MODA/inst/doc/MODA.html)

